example.py

```diff
@@ -6,29 +6,44 @@
 
 
 class Producer(threading.Thread):
-    daemon = True
+    def __init__(self):
+        threading.Thread.__init__(self)
+        self.stop_event = threading.Event()
+        
+    def stop(self):
+        self.stop_event.set()
 
     def run(self):
         producer = KafkaProducer(bootstrap_servers='localhost:9092')
 
-        while True:
+        while not self.stop_event.is_set():
             producer.send('my-topic', b"test")
             producer.send('my-topic', b"\xc2Hola, mundo!")
             time.sleep(1)
 
+        producer.close()
 
 class Consumer(multiprocessing.Process):
-    daemon = True
-
+    def __init__(self):
+        multiprocessing.Process.__init__(self)
+        self.stop_event = multiprocessing.Event()
+        
+    def stop(self):
+        self.stop_event.set()
+        
     def run(self):
         consumer = KafkaConsumer(bootstrap_servers='localhost:9092',
                                  auto_offset_reset='earliest')
         consumer.subscribe(['my-topic'])
 
-        for message in consumer:
-            print (message)
-
+        while not self.stop_event.is_set():
+            message = consumer.poll(timeout_ms=0, max_records=1)
```

Comment:
I admit that I am not a huge fan of the `poll` interface. We support it because that is how the java client works, but I think using the iterator is much more pythonic. Perhaps something like:
```
consumer = KafkaConsumer(...., consumer_timeout_ms=1000)
while True:
    for message in consumer:
        print(message)
        if self.stop_event.is_set():
            break
````

?

Reply:
It will not work. Needs the wakeup method

Reply:
I agree the iterator looks more pythonic, and believe me I've tried to use it instead. But it simply isn't controllable as currently implemented. The main problem being that it blocks indefinitely if it's called when there won't be any new messages for it. 

consumer_timeout_ms is not a great solution either because it just statically sets the lifetime of the KafkaConsumer object. The consumer just closes on its own after that length of time even if you're actively pulling messages through it with the iterator. You can set consumer_timeout_ms for longer, but then you're just back to the same problem where the iterator call is blocking forever until that timer runs out. 

poll() is the only way I've found to put clean consumer closure under programmer control. It lets you use the same consumer as long as you like without reinitializing it, and then lets you shut it down very quickly as well. My code, which poll()'s for maxrecords=1 and loops if no new messages, is the closest you can currently get to a controllable iterator that allows the programmer's code to end the request loop at will.

If I've somehow missed a command to shut down the iterator's internal request loop please let me know.